migrator/zenodo_rdm_migrator/extract/kafka.py

```diff
@@ -7,33 +7,348 @@
+        if ts:
+            offsets = consumer.offsets_for_times({p: ts for p in partitions})
+            partitions.update({p: o for p, (o, _) in offsets.items()})
+        else:
+            # ...else start from the very beginning of the topic partitions
+            partitions = consumer.beginning_offsets(partitions)
+
+        for partition, offset in partitions.items():
+            consumer.seek(partition, offset)
+        return partitions
+
+    def _seek_committed_offsets(self, consumer, topic):
+        partitions = [
+            TopicPartition(topic, p) for p in consumer.partitions_for_topic(topic)
+        ]
+        offsets = [consumer.committed(p) for p in partitions]
+        consumer.assign(partitions)
+        for partition, offset in zip(partitions, offsets):
+            consumer.seek(partition, offset)
+
+    def _get_consumer(self, topic, group_id):
+        consumer = KafkaConsumer(
+            group_id=group_id,
+            **self.DEFAULT_CONSUMER_CFG,
+            **self.config,
+        )
+        if topic not in self._topic_states:
+            self._topic_states[topic] = self._seek_dt_offsets(
+                consumer,
+                topic,
+                ts=self.from_ts,
+            )
+        else:
+            self._seek_committed_offsets(consumer, topic)
+        return consumer
+
+    # NOTE: These two properties are useful for tests/mocking
+    @property
+    def _tx_consumer(self):
+        return self._get_consumer(self.tx_topic, "zenodo_migration_tx")
+
+    @property
+    def _ops_consumer(self):
+        return self._get_consumer(self.ops_topic, "zenodo_migration_ops")
+
+    def iter_tx_info(self):
+        """Yield commited transactions info."""
+        consumer = self._tx_consumer
+        for tx_msg in consumer:
+            self._dump_msg(self.tx_topic, tx_msg)
+
+            if not tx_msg.value:
+                # Sometimes messages don't contain a value... So far it's not been an
+                # issue, but maybe logging in DEBUG could help at some point.
+                self.logger.debug(f"No message value for tx_info {tx_msg}")
+                consumer.commit()
+                continue
+
+            tx_id, lsn = map(int, tx_msg.value["id"].split(":"))
+            # We drop anything before the configured last transaction ID
+            if tx_id <= self.last_tx:
+                consumer.commit()
+                continue
+            if tx_msg.value["status"] == "BEGIN":
+                # ignore BEGIN statements
+                consumer.commit()
+                continue
+            elif tx_msg.value["status"] == "END":
+                consumer.commit()
+                yield (tx_id, tx_msg.value)
+
+    @staticmethod
+    def _filter_unchanged_values(msg):
+        # We don't have deeply nested dict values, since they are still raw JSON strings
+        before = msg.value["before"]
+        after = msg.value["after"]
+        pk_keys = set(msg.key.keys())
```

Comment:
self note: I assume this `.key` with the pks is something that kafka/debezium gives? I see it's being extracted from `op_msg.key` coming from the consumer messages.